File: internal/signal/surrogate.go
// Package signal 提供确定性的 LLM/embedder 替身信号源（离线评测用）。
//
// 两种替身（spec/03-oracles.md §3 的评测 oracle 数据源）：
//   - Surrogate（本文件）：从画像文本计算语义可解释的确定性信号
//     （token 重叠式方向分 / TF 余弦 embedding 分 + 固定 seed 噪声），
//     模拟真实 LLM/embedding 的判断质量；
//   - FakeLLM / FakeEmbedder（fake.go）：按 prompt 查表返回固定响应，
//     守护 stage 契约（spec/04-fixtures.md §7）。
//
// 与 Python 基线（src/mutual/surrogate.py）逐位对齐：随机噪声流
// RandomState(seed).rand() 的消费顺序 = member × pool × (a_to_b, b_to_a)。
package signal

import (
	"math"
	"regexp"
	"sort"

	"github.com/Cloudbird-Software/mutual/internal/num"
	"github.com/Cloudbird-Software/mutual/internal/rng"
)

// 与 config/default.yaml recipe.section_weights 对齐的代理权重。
const (
	wNeedsSkills   = 0.6
	wProjectSkills = 0.2
	wVision        = 0.2
)

// sectionJoinOrder 是画像级 embedding 拼接的分节顺序（Python _SECTION_KEYS）。
var sectionJoinOrder = []string{"needs", "project", "skills", "vision"}

var wordRE = regexp.MustCompile(`[a-z0-9]+`)

// Tokenize 小写化并切词：ASCII 走 [a-z0-9]+ 词切分（与 Python 基线
// 逐位一致，golden 语义不变）；CJK 字符追加二元组（无分词器的廉价
// 语义近似——海外商人↔本地企业等跨语言画像的离线评测可观测性，
// 2026-08 实验 R4：纯 ASCII 切词下中文侧词法全盲，NSW 几何均值
// 湮灭黄金对，LLM 行级 10/10 全解）。
func Tokenize(text string) []string {
	t := lower(text)
	toks := wordRE.FindAllString(t, -1)
	toks = append(toks, cjkBigrams(t)...)
	return toks
}

// cjkBigrams 提取 CJK 统一表意文字区段的字符二元组（单字原样保留）。
// 扩展区（ExtA/Compat/B/C/D）与日文假名不覆盖：合成语料以简中为主，
// 按需扩展。
func cjkBigrams(text string) []string {
	var out []string
	runes := []rune(text)
	i := 0
	for i < len(runes) {
		if !isHan(runes[i]) {
			i++
			continue
		}
		j := i
		for j < len(runes) && isHan(runes[j]) {
			j++
		}
		for k := i; k < j; k++ {
			if k+1 < j {
				out = append(out, string(runes[k:k+2]))
			} else if k == i {
				// 单字 run：保留单字（可与其他 run 的二元组区分开）
				out = append(out, string(runes[k]))
			}
		}
		i = j
	}
	return out
}

func isHan(r rune) bool {
	return (r >= 0x4E00 && r <= 0x9FFF) || // CJK 统一表意文字
		(r >= 0x3400 && r <= 0x4DBF) // 扩展 A
}

func lower(s string) string {
	// ASCII 快路径覆盖画像文本的约定字符集；非 ASCII 字符按 rune 小写化。
	needRune := false
	for i := 0; i < len(s); i++ {
		if s[i] >= 0x80 {
			needRune = true
			break
		}
	}
	if !needRune {
		b := []byte(s)
		for i, c := range b {
			if c >= 'A' && c <= 'Z' {
				b[i] = c + ('a' - 'A')
			}
		}
		return string(b)
	}
	runes := []rune(s)
	for i, r := range runes {
		runes[i] = runeDown(r)
	}
	return string(runes)
}

func runeDown(r rune) rune {
	if r >= 'A' && r <= 'Z' {
		return r + ('a' - 'A')
	}
	return r
}

// tokenSet 提取某分节的 token 集合（缺失分节 → 空集）。
func tokenSet(sections map[string]string, key string) map[string]bool {
	out := map[string]bool{}
	for _, tok := range Tokenize(sections[key]) {
		out[tok] = true
	}
	return out
}

// overlap 计算集合重叠的余弦式度量 ∈ [0, 1]；双方皆空记 0（中性偏保守）。
func overlap(a, b map[string]bool) float64 {
	if len(a) == 0 || len(b) == 0 {
		return 0
	}
	inter := 0
	for tok := range a {
		if b[tok] {
			inter++
		}
	}
	return float64(inter) / math.Sqrt(float64(len(a)*len(b)))
}

// DirectionalScore 返回 A 对 B 的方向性价值分 ∈ [0, 1]
// （模拟 LLM 的 a_to_b 打分，与 recipe.instruction 语义对齐）：
//
//	0.6 · overlap(A.needs,   B.skills)  —— 需求被技能直击（主信号）
//	0.2 · overlap(A.project, B.skills)  —— 项目协作空间
//	0.2 · overlap(A.vision,  B.vision)  —— 方向一致
func DirectionalScore(sectionsA, sectionsB map[string]string) float64 {
	return wNeedsSkills*overlap(tokenSet(sectionsA, "needs"), tokenSet(sectionsB, "skills")) +
		wProjectSkills*overlap(tokenSet(sectionsA, "project"), tokenSet(sectionsB, "skills")) +
		wVision*overlap(tokenSet(sectionsA, "vision"), tokenSet(sectionsB, "vision"))
}

// tfVectors 构建固定词表 TF 向量并 L2 归一化（确定性，无外部模型）。
//
// 词表按文本顺序的首次出现构建；归一化走 num.Norm（NumPy pairwise
// 求和，逐位对齐 np.linalg.norm）。
func tfVectors(texts []string) [][]float64 {
	vocab := map[string]int{}
	tokenLists := make([][]string, len(texts))
	for r, t := range texts {
		tokenLists[r] = Tokenize(t)
		for _, tok := range tokenLists[r] {
			if _, ok := vocab[tok]; !ok {
				vocab[tok] = len(vocab)
			}
		}
	}
	mat := make([][]float64, len(texts))
	for r := range mat {
		mat[r] = make([]float64, len(vocab))
	}
	for r, toks := range tokenLists {
		for _, tok := range toks {
			mat[r][vocab[tok]]++
		}
	}
	for _, row := range mat {
		norm := num.Norm(row)
		if norm > 1e-12 {
			for d := range row {
				row[d] /= norm
			}
		}
	}
	return mat
}

// EmbedScore 返回画像级 embedding 相似度 ∈ [0, 1]
// （四个分节按固定顺序拼接后 TF 余弦，截断到 [0, 1]）。
func EmbedScore(sectionsA, sectionsB map[string]string) float64 {
	mat := tfVectors([]string{joinSections(sectionsA), joinSections(sectionsB)})
	if len(mat[0]) == 0 {
		return 0
	}
	dot := 0.0
	for d := range mat[0] {
		dot += mat[0][d] * mat[1][d]
	}
	return math.Max(0, math.Min(1, dot))
}

func joinSections(sections map[string]string) string {
	parts := make([]string, 0, len(sectionJoinOrder))
	for _, k := range sectionJoinOrder {
		parts = append(parts, sections[k])
	}
	out := ""
	for i, p := range parts {
		if i > 0 {
			out += " "
		}
		out += p
	}
	return out
}

// Noisy 加性噪声并截断到 [0, 1]（固定 seed → 确定性）。
//
// 每次调用消费 RandomState 的一个均匀随机数（与 Python
// np.clip(score + scale*(rng.rand()-0.5), 0, 1) 逐位一致）。
func Noisy(score float64, rs *rng.RandomState, noiseScale float64) float64 {
	return math.Max(0, math.Min(1, score+noiseScale*(rs.Float64()-0.5)))
}

// DirScore 是一对双向分数（a_to_b, b_to_a）。
type DirScore struct {
	AToB float64
	BToA float64
}

// OrderedSections 是保序的分节数据（map 迭代无序，评测链路的噪声
// 流依赖 member/pool 的文件顺序，故用显式顺序结构）。
type OrderedSections struct {
	ID       string
	Sections map[string]string
}

// ScoreMatrix 批量计算双向偏好分（模拟 score 阶段输出，喂给 pre_matrix）。
//
// 随机流消费顺序（与 Python 逐位一致）：member 外层 × pool 内层，
// 每对依次 noisy(a_to_b)、noisy(b_to_a)。embeddingOnly 时退化为纯
// embedding 相似度（冷启动：无打分历史，双向同值）。
//
// 调用方必须传入保序的 members/pool（JSON 文件序）。
func ScoreMatrix(
	members, pool []OrderedSections,
	seed int,
	noiseScale float64,
	embeddingOnly bool,
) map[string]map[string]DirScore {
	rs := rng.New(uint32(mod32(seed)))
	out := map[string]map[string]DirScore{}
	for _, m := range members {
		row := map[string]DirScore{}
		for _, p := range pool {
			if embeddingOnly {
				s := EmbedScore(m.Sections, p.Sections)
				row[p.ID] = DirScore{AToB: Noisy(s, rs, noiseScale), BToA: Noisy(s, rs, noiseScale)}
			} else {
				a2b := DirectionalScore(m.Sections, p.Sections)
				b2a := DirectionalScore(p.Sections, m.Sections)
				row[p.ID] = DirScore{AToB: Noisy(a2b, rs, noiseScale), BToA: Noisy(b2a, rs, noiseScale)}
			}
		}
		out[m.ID] = row
	}
	return out
}

// ScoreMatrixBlended 是真实管线 score 阶段双信号混合的离线镜像：
// pref = embedW·noisy(EmbedScore) + llmW·noisy(DirectionalScore)，双向独立。
//
// 用途：官方 bench 的 ScoreMatrix 只喂方向分（LLM 信号替身），完全绕过
// config blending（embed/llm 权重）——真实管线里 embed 相似度是同义改写
// 盲区的兜底信号（classic m3↔p3 类失效）。本函数让离线评测能执行与
// config/default.yaml 相同的混合语义。
//
// 噪声流：方向分与 ScoreMatrix 逐位一致（member × pool × (a_to_b, b_to_a)）；
// embed 分走独立流（seed+777777，同序）。embedW=0 时输出与 ScoreMatrix
// 完全一致（零值兼容，golden 不受影响）。embeddingOnly 场景不适用
// （冷启动无 LLM 信号可混合），调用方负责分流。
func ScoreMatrixBlended(
	members, pool []OrderedSections,
	seed int,
	noiseScale, embedW, llmW float64,
) map[string]map[string]DirScore {
	out := ScoreMatrix(members, pool, seed, noiseScale, false)
	if embedW <= 0 {
		return out
	}
	rs := rng.New(uint32(uint32(mod32(seed) + 777777)))
	for _, m := range members {
		for _, p := range pool {
			en := Noisy(EmbedScore(m.Sections, p.Sections), rs, noiseScale)
			s := out[m.ID][p.ID]
			out[m.ID][p.ID] = DirScore{
				AToB: clamp01f(embedW*en + llmW*s.AToB),
				BToA: clamp01f(embedW*en + llmW*s.BToA),
			}
		}
	}
	return out
}

func clamp01f(v float64) float64 {
	return math.Max(0, math.Min(1, v))
}

func mod32(seed int) int {
	u := uint64(uint32(seed))
	return int(u)
}

// SortedKeys 返回 map 键的排序副本（确定性遍历 helper，测试用）。
func SortedKeys[V any](m map[string]V) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}

File: src/mutual/surrogate.py
"""Mutual — 确定性代理信号源（离线评测用 LLM/embedder 替身）。

Phase 3 评测闭环（docs/engineering-plan.md §5.1）在 CI 中必须离线可复现，
但真实 LLM 打分不可复现且需凭据。本模块提供**语义可解释的确定性替代**：

- :func:`directional_score`：needs↔skills / project↔skills / vision 重叠
  （token 余弦式），模拟 LLM 双向打分的语义判断；
- :func:`embed_score`：TF 向量余弦，模拟 embedding 相似度（冷启动场景的
  唯一信号——无打分历史时系统只能依赖 embedding）；
- 噪声项 ``noise_scale``：固定 seed 的加性噪声，模拟 LLM 判断的不完美
  （无噪声时打分完全等同于 ground truth 构造规则 → HR 恒 1.0，门禁失去
  判别力；有噪声时求解器/排序的退化会真实传导到 HR/NDCG）。

与 fake_llm（spec/04-fixtures.md §7）的区别：fake_llm 按 prompt 查表返回
固定响应（守护 stage 契约）；surrogate 从画像文本计算信号（评测语义）。
"""

from __future__ import annotations

import re
from typing import Dict, List, Sequence

import numpy as np

# 与 config/default.yaml recipe.section_weights 对齐的代理权重
_W_NEEDS_SKILLS = 0.6
_W_PROJECT_SKILLS = 0.2
_W_VISION = 0.2

_WORD_RE = re.compile(r"[a-z0-9]+")

_SECTION_KEYS = ("needs", "project", "skills", "vision")


def tokenize(text: str) -> List[str]:
    """小写化并按非字母数字切词（画像文本约定为英文关键词风格）。"""
    return _WORD_RE.findall(text.lower())


def _tokens_of(sections: Dict[str, str], key: str) -> set[str]:
    return set(tokenize(sections.get(key) or ""))


def _overlap(a: set[str], b: set[str]) -> float:
    """集合重叠的余弦式度量 ∈ [0, 1]；双方皆空记 0（中性偏保守）。"""
    if not a or not b:
        return 0.0
    return len(a & b) / np.sqrt(len(a) * len(b))


def directional_score(
    sections_a: Dict[str, str],
    sections_b: Dict[str, str],
) -> float:
    """A 对 B 的方向性价值分 ∈ [0, 1]（模拟 LLM 的 a_to_b 打分）。

    语义规则（与 recipe.instruction 对齐：A 的需求被 B 的技能满足 +
    共享方向）：
    - 0.6 * overlap(A.needs, B.skills)   —— 需求被技能直击（主信号）
    - 0.2 * overlap(A.project, B.skills) —— 项目协作空间
    - 0.2 * overlap(A.vision, B.vision)  —— 方向一致
    """
    return float(
        _W_NEEDS_SKILLS
        * _overlap(_tokens_of(sections_a, "needs"), _tokens_of(sections_b, "skills"))
        + _W_PROJECT_SKILLS
        * _overlap(_tokens_of(sections_a, "project"), _tokens_of(sections_b, "skills"))
        + _W_VISION * _overlap(_tokens_of(sections_a, "vision"), _tokens_of(sections_b, "vision"))
    )


def tf_vector(texts: Sequence[str]) -> np.ndarray:
    """构建固定词表 TF 向量矩阵（确定性，无外部模型）。"""
    vocab: Dict[str, int] = {}
    token_lists = [tokenize(t) for t in texts]
    for toks in token_lists:
        for tok in toks:
            if tok not in vocab:
                vocab[tok] = len(vocab)
    mat = np.zeros((len(texts), len(vocab)), dtype=float)
    for r, toks in enumerate(token_lists):
        for tok in toks:
            mat[r, vocab[tok]] += 1.0
    # L2 归一化 → 点积即余弦
    norms = np.linalg.norm(mat, axis=1, keepdims=True)
    norms[norms == 0] = 1.0
    return mat / norms


def embed_score(
    sections_a: Dict[str, str],
    sections_b: Dict[str, str],
) -> float:
    """画像级 embedding 相似度 ∈ [0, 1]（四个 section 拼接后 TF 余弦）。"""
    mat = tf_vector([_join_sections(sections_a), _join_sections(sections_b)])
    return float(np.clip(np.dot(mat[0], mat[1]), 0.0, 1.0))


def _join_sections(sections: Dict[str, str]) -> str:
    return " ".join((sections.get(k) or "") for k in _SECTION_KEYS)


def noisy(score: float, rng: np.random.RandomState, noise_scale: float) -> float:
    """加性噪声并截断到 [0, 1]（固定 seed → 确定性）。"""
    return float(np.clip(score + noise_scale * (rng.rand() - 0.5), 0.0, 1.0))


def score_matrix(
    member_sections: Dict[str, Dict[str, str]],
    pool_sections: Dict[str, Dict[str, str]],
    seed: int,
    noise_scale: float = 0.15,
    embedding_only: bool = False,
) -> Dict[str, Dict[str, tuple[float, float]]]:
    """批量计算双向偏好分（模拟 score 阶段输出，喂给 pre_matrix）。

    Args:
        member_sections: ``{member_id → {section → text}}``。
        pool_sections: ``{pool_id → {section → text}}``。
        seed: 噪声随机种子（确定性）。
        noise_scale: 噪声幅度；0 = 完美信号（调试用）。
        embedding_only: True 时退化为纯 embedding 相似度（冷启动：无打分
            历史，双向同值）。

    Returns:
        ``{member_id → {pool_id → (a_to_b, b_to_a)}}``。
    """
    rng = np.random.RandomState(seed)
    out: Dict[str, Dict[str, tuple[float, float]]] = {}
    for mid, msec in member_sections.items():
        row: Dict[str, tuple[float, float]] = {}
        for pid, psec in pool_sections.items():
            if embedding_only:
                s = embed_score(msec, psec)
                row[pid] = (noisy(s, rng, noise_scale), noisy(s, rng, noise_scale))
            else:
                a2b = directional_score(msec, psec)
                b2a = directional_score(psec, msec)
                row[pid] = (noisy(a2b, rng, noise_scale), noisy(b2a, rng, noise_scale))
        out[mid] = row
    return out
